client/my-sites/checkout/checkout/index.jsx

```diff
@@ -394,8 +396,34 @@ export class Checkout extends React.Component {
 		} = this.props;
 		const domainReceiptId = get( getGoogleApps( cart ), '[0].extra.receipt_for_domain', 0 );
 
+		const adminUrl = get( selectedSite, [ 'options', 'admin_url' ] );
+
+		// If we're given an explicit `redirectTo` query arg, make sure it's either internal
+		// (i.e. on WordPress.com), or a Jetpack or WP.com site's block editor (in wp-admin).
+		// This is required for Jetpack's (and WP.com's) paid blocks Upgrade Nudge.
 		if ( redirectTo ) {
-			return redirectTo;
+			if ( ! isExternal( redirectTo ) ) {
+				return redirectTo;
+			}
+
+			const { protocol, hostname, port, pathname, query } = parseUrl( redirectTo, true, true );
+
+			// We cannot simply compare `hostname` to `selectedSiteSlug`, since the latter
+			// might contain a path in the case of Jetpack subdirectory installs.
+			if ( redirectTo.startsWith( adminUrl ) && pathname.endsWith( '/wp-admin/post.php' ) ) {
```

Comment:
As written, this would be OK with the `adminUrl` get above that's potentially `undefined`:

```js
'anything'.startsWith( undefined ) // false
```